src/verifier.rs

```diff
@@ -0,0 +1,1262 @@
+use std::{collections::HashMap, fs::File, io::copy, path::Path, str::FromStr};
+
+use ark_bn254::{Fq, G1Affine};
+use bytes::Bytes;
+use ethabi::{encode, Token};
+use ethereum_types::{Address, U256};
+use rust_kzg_bn254::{blob::Blob, kzg::Kzg, polynomial::PolynomialFormat};
+use tiny_keccak::{Hasher, Keccak};
+
+use super::{
+    blob_info::{BatchHeader, BlobHeader, BlobInfo, G1Commitment},
+    config::PointsSource,
+    eth_client::{EthClient, EthClientError},
+    BATCH_ID_TO_METADATA_HASH_FUNCTION_SELECTOR,
+    QUORUM_ADVERSARY_THRESHOLD_PERCENTAGES_FUNCTION_SELECTOR,
+    QUORUM_NUMBERS_REQUIRED_FUNCTION_SELECTOR,
+};
+
+#[derive(Debug)]
+pub enum VerificationError {
+    ServiceManagerError,
+    KzgError,
+    WrongProof,
+    DifferentCommitments,
+    DifferentRoots,
+    EmptyHash,
+    DifferentHashes,
+    WrongQuorumParams,
+    QuorumNotConfirmed,
+    CommitmentNotOnCurve,
+    CommitmentNotOnCorrectSubgroup,
+    LinkError,
+}
+
+#[async_trait::async_trait]
+pub trait VerifierClient: Sync + Send + std::fmt::Debug {
+    fn clone_boxed(&self) -> Box<dyn VerifierClient>;
+
+    /// Returns the current block number.
+    async fn get_block_number(&self) -> Result<U256, EthClientError>;
+
+    /// Invokes a function on a contract specified by `contract_address` / `contract_abi` using `eth_call`.
+    async fn call(
+        &self,
+        to: Address,
+        calldata: Bytes,
+        block: Option<u64>,
+    ) -> Result<String, EthClientError>;
+}
+
+#[async_trait::async_trait]
+impl VerifierClient for EthClient {
+    fn clone_boxed(&self) -> Box<dyn VerifierClient> {
+        Box::new(self.clone())
+    }
+
+    async fn get_block_number(&self) -> Result<U256, EthClientError> {
+        self.get_block_number().await
+    }
+
+    async fn call(
+        &self,
+        to: Address,
+        calldata: Bytes,
+        block: Option<u64>,
+    ) -> Result<String, EthClientError> {
+        self.call(to, calldata, block).await
+    }
+}
+
+/// Configuration for the verifier used for authenticated dispersals
+#[derive(Debug, Clone)]
+pub struct VerifierConfig {
+    pub svc_manager_addr: String,
+    pub max_blob_size: u32,
+    pub points: PointsSource,
+    pub settlement_layer_confirmation_depth: u32,
+}
+
+/// Verifier used to verify the integrity of the blob info
+/// Kzg is used for commitment verification
+/// EigenDA service manager is used to connect to the service manager contract
+#[derive(Debug)]
+pub struct Verifier {
+    kzg: Kzg,
+    cfg: VerifierConfig,
+    eth_client: Box<dyn VerifierClient>,
+}
+
+impl Clone for Verifier {
+    fn clone(&self) -> Self {
+        Self {
+            kzg: self.kzg.clone(),
+            cfg: self.cfg.clone(),
+            eth_client: self.eth_client.clone_boxed(),
+        }
+    }
+}
+
+impl Verifier {
+    async fn save_points(link: String) -> Result<String, VerificationError> {
```

Comment:
rename to download_points? save sounds like its just saving from ram to storage.

Reply:
Done